http/client.go

```diff
@@ -144,53 +152,47 @@ func New(options ...Option) *Client {
 	}
 
 	return &Client{
-		Client:          *http.DefaultClient,
+		Client:          *opts.httpClient,
 		headersFunc:     opts.headersFunc,
 		maxSendAttempts: MaxSendAttempts,
 	}
 }
 
 // NewNonSSLValidating creates a new goose http *Client skipping SSL validation.
+// DEPRECATED (stickupkid): No longer used internally, left to ensure we don't
+// break compatibility.
 func NewNonSSLValidating(options ...Option) *Client {
-	opts := newOptions()
-	for _, option := range options {
-		option(opts)
-	}
-
-	insecureClientMutex.Lock()
-	httpClient := insecureClient
-	if httpClient == nil {
-		insecureConfig := &tls.Config{InsecureSkipVerify: true}
-		insecureTransport := &http.Transport{TLSClientConfig: insecureConfig}
-		insecureClient = &http.Client{Transport: insecureTransport}
-		httpClient = insecureClient
-	}
-	insecureClientMutex.Unlock()
-
-	return &Client{
-		Client:          *httpClient,
-		headersFunc:     opts.headersFunc,
-		maxSendAttempts: MaxSendAttempts,
-	}
+	return NewWithTLSConfig(&tls.Config{
+		InsecureSkipVerify: true,
+	}, options...)
 }
 
 // NewWithTLSConfig creates a new goose http *Client with a TLS config.
+// DEPRECATED (stickupkid): No longer used internally, left to ensure we don't
+// break compatibility.
 func NewWithTLSConfig(tlsConfig *tls.Config, options ...Option) *Client {
 	opts := newOptions()
 	for _, option := range options {
 		option(opts)
 	}
 
-	defaultClient := *http.DefaultClient
-	defaultClient.Transport = &http.Transport{
-		TLSClientConfig: tlsConfig,
+	// Ensure we don't overwrite existing http.Client information.
+	client := *opts.httpClient
+	if opts.httpClient == nil {
+		client = http.Client{}
 	}
-
-	return &Client{
-		Client:          defaultClient,
-		headersFunc:     opts.headersFunc,
-		maxSendAttempts: MaxSendAttempts,
+	if client.Transport == nil {
+		client.Transport = &http.Transport{}
 	}
+	if t, ok := client.Transport.(*http.Transport); ok {
```

Comment:
We should return an error if ok false. Otherwise you'd get a client without a TLS config and not know it.